src/core/sock/sockinfo.h

```diff
@@ -644,4 +649,80 @@ class sockinfo : public socket_fd_api,
     //////////////////////////////////////////////////////////////////
 };
 
+typedef std::unordered_map<flow_tuple, bool> port_no_bind_map_t;
+struct port_map_list {
+    std::list<in_port_t>::iterator it;
+    port_no_bind_map_t map;
+};
+
+class bind_no_port {
+public:
+    /*  Logic of IP_BIND_ADDRESS_NO_PORT
+        In case we need a new port from OS, we call bind with port 0,
+        Otherwise - we call bind with specific port from our DB. */
+    int bind_and_set_port_map(const sock_addr &src, const sock_addr &dst, int fd)
+    {
+        int ret = 0;
+        sock_addr addr(src);
+        socklen_t addr_len = sizeof(addr);
+        std::lock_guard<decltype(m_lock)> lock(m_lock);
+
+        in_port_t reused_port = 0;
+        flow_tuple temp_mapper(dst.get_ip_addr(), dst.get_in_port(), src.get_ip_addr(),
+                               src.get_in_port(), PROTO_TCP, src.get_sa_family());
+        for (std::list<in_port_t>::iterator _port_it = m_port_list.begin();
+             _port_it != m_port_list.end(); ++_port_it) {
+            auto _iter = arr[*_port_it].map.find(temp_mapper);
+            if (_iter == arr[*_port_it].map.end()) {
+                reused_port = *_port_it;
+                arr[*_port_it].it = _port_it;
+                break;
+            }
+        }
+
+        if (reused_port) {
+            addr.set_in_port(reused_port);
+            arr[reused_port].map[temp_mapper] = true;
+        }
+
+        if ((ret = orig_os_api.bind(fd, addr.get_p_sa(), addr_len))) {
```

Comment:
What will happen in this bind if this port already bound?

Reply:
default call is with port=0.
the only way we will bind to specific port - if the port is already in our DB, and for that we had to use so_reuseport, for both previous socket, and the current socket. So bind shouldn't fail anyway.